lib/src/main/java/edu/stanford/futuredata/macrobase/analysis/summary/IncrementalSummarizer.java

```diff
@@ -0,0 +1,320 @@
+            outlierItemsetPaneCounts = new ArrayDeque<>(numPanes);
+        }
+    }
+
+    @Override
+    public int getWindowSize() { return numPanes; }
+
+    public IncrementalSummarizer setMinSupport(double minSupport) {
+        this.minOutlierSupport = minSupport;
+        return this;
+    }
+
+    /**
+     * Adjust this to tune the severity (e.g. strength of correlation) of the results returned.
+     * @param minRiskRatio lowest risk ratio to consider for meaningful explanations.
+     * @return this
+     */
+    public IncrementalSummarizer setMinRiskRatio(double minRiskRatio) {
+        this.minRiskRatio = minRiskRatio;
+        return this;
+    }
+
+    /**
+     * By default, will check for nonzero entries in a column of doubles.
+     * @param predicate function to signify whether row should be treated as outlier.
+     * @return this
+     */
+    public IncrementalSummarizer setOutlierPredicate(DoublePredicate predicate) {
+        this.predicate = predicate;
+        return this;
+    }
+    public IncrementalSummarizer setAttributes(List<String> attributes) {
+        this.attributes = attributes;
+        this.encoder.setColumnNames(attributes);
+        return this;
+    }
+
+    /**
+     * Whether or not to use combinations of attributes in explanation, or only
+     * use simple single attribute explanations
+     * @param useAttributeCombinations flag
+     * @return this
+     */
+    public IncrementalSummarizer setUseAttributeCombinations(boolean useAttributeCombinations) {
+        this.useAttributeCombinations = useAttributeCombinations;
+        return this;
+    }
+
+    /**
+     * Set the column which indicates outlier status. "_OUTLIER" by default.
+     * @param outlierColumn new outlier indicator column.
+     * @return this
+     */
+    public IncrementalSummarizer setOutlierColumn(String outlierColumn) {
+        this.outlierColumn = outlierColumn;
+        return this;
+    }
+
+    protected void encodeAttributes(DataFrame df) {
+        // Filter inliers and outliers
+        DataFrame outlierDF = df.filter(outlierColumn, predicate);
+        DataFrame inlierDF = df.filter(outlierColumn, predicate.negate());
+
+        // Encode inlier and outlier attribute columns
+        if (attributes.isEmpty()) {
+            encoder.setColumnNames(df.getSchema().getColumnNamesByType(Schema.ColType.STRING));
+            inlierItemsets = encoder.encodeAttributes(inlierDF.getStringCols());
+            outlierItemsets = encoder.encodeAttributes(outlierDF.getStringCols());
+        } else {
+            encoder.setColumnNames(attributes);
+            inlierItemsets = encoder.encodeAttributes(inlierDF.getStringColsByName(attributes));
+            outlierItemsets = encoder.encodeAttributes(outlierDF.getStringColsByName(attributes));
+        }
+        inlierPaneCounts.add(inlierItemsets.size());
+        outlierPaneCounts.add(outlierItemsets.size());
+    }
+
+    private void calcCumSum() {
```

Comment:
add comment for this function... what is it doing? where is it called? why is it protected?